torch/_inductor/lowering.py

```diff
@@ -6219,10 +6220,199 @@ def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs):
+                gm=subgraph.graph_module,
+                example_inputs=fx_all_inputs,  # type: ignore[arg-type]
+                subgraph_name=subgraph.name,
+            )
+            with V.set_graph_handler(subgraph.graph):
+                subgraph.graph.run(*fake_all_inputs)
+
     result = ir.WhileLoop.create(cond_fn, body_fn, carried_inputs, additional_inputs)
     return list(map(TensorBox.create, result))
 
 
+# This is a spcialized version of select because we're certain that the idx is valid
+@register_lowering(torch.ops._scan_helper.unsafe_select, type_promotion_kind=None)
+def _(dst, dim, idx):
+    return squeeze(slice_(dst, dim, idx, idx + 1, step=1, clamp=False), dim)
+
+
+@register_lowering(scan_op)
+def scan(combine_subgraph, init, xs, dim, reverse, additional_inputs):
+    from torch._dynamo.source import GlobalSource
+    from torch._higher_order_ops.scan import _extract_carry_and_out, stack_y
+    from torch._inductor.pattern_matcher import fwd_only
+    from torch.fx.experimental.symbolic_shapes import DimDynamic
+
+    if torch._inductor.config.cpp_wrapper:
+        raise NotImplementedError("scan is not supported with cpp_wrapper yet.")
+
+    num_init_leaves = len(init)
+    specialized_dim = int(dim)
+
+    with V.graph.fake_mode:
+        # We choose tensor box to be the single source of truth of creating fake.
+        # tensors for subgraph.
+        # It's not reliable to get fake tensor from the "val" meta of fx node.
+        # For example: the View ir node in inductor will create a new TensorBox
+        # whose size is exactly the viewed shape but in fake tensor
+        # propagation, we it's not guaranteed to be the case. Check the decomposition
+        # of view operator.
+        def fake_tensor_from_tbox(tbox):
+            tbox.realize()
+            with V.graph.fake_mode:
+                shape_env = V.graph.fake_mode.shape_env
+
+                def _find_all_symints(fake: Union[torch.SymInt, torch.Tensor]):
+                    assert isinstance(fake, (torch.Tensor, torch.SymInt))
+                    if isinstance(fake, torch.Tensor):
+                        for sz in fake.size():
+                            if isinstance(sz, torch.SymInt):
+                                all_symints.update({str(sz): sz})
+                        for st in fake.size():
+                            if isinstance(st, torch.SymInt):
+                                all_symints.update({str(st): st})
+                    elif isinstance(fake, torch.SymInt):
+                        all_symints.update({str(fake): fake})
+
+                all_symints: Dict[str, torch.SymInt] = {}
+                pytree.tree_map(
+                    lambda fake: _find_all_symints(fake.fake), shape_env.tracked_fakes  # type: ignore[union-attr]
+                )
+                fake_size = [eval(str(sz), all_symints) for sz in tbox.get_size()]
+                fake_stride = [eval(str(sz), all_symints) for sz in tbox.get_stride()]
```

Comment:
eval is sus here - you should not need to do this. you can use `_SymHashingDict` as you need it.

Reply:
Switching to meta["val"] avoids this.